### PR TITLE
ENH: add regex parameter to .loc

### LINKED ISSUE #26642: Rename NDFrame.filter to .select?
I've made a PR to remove ``NDFrame.select``, see #26641. That method was deprecated in 0.21.

There was a discussion in #12401, where many participants agreed that removing the old ``select`` and having ``NDFrame.filter`` renamed to ``NDFrame.select`` would be a better naming scheme. I agree on that for several of the reasons mentioned in the thread.

What do people think of such a name change and a related deprecation of the name (not functionality) ``.filter``?

### PR DESCRIPTION
**Pull request by topper-123**
- [x] closes #26642
- [x] xref #27340
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Adds a ``regex`` parameter to ``NDFrame.loc.__call__``, so we can now do:

```python
>>> df = pd.DataFrame(1, index=["A", "AB", "BC", "CD"], columns=["CD", "BC", "AB", "A"])
>>> df.loc(regex=True)["B", "B"]
    BC  AB
AB   1   1
BC   1   1
```

This is intended to be followed up by deprecating ``NDFrame.filter``.

This implementation does not allow setting values using ``regex=True``, as the most urgent motivation is to get a replacement for ``NDFrame.filter``. Setting values can come later, even after version 1.0.

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -1822,10 +1855,21 @@ def _getitem_axis(self, key, axis: int):
                 indexer[axis] = locs
                 return self.obj.iloc[tuple(indexer)]
 
+        if self.regex and isinstance(key, str):
```
Comment (simonjayhawkins) [inline pandas/core/indexing.py]: would we want to raise if regex and key is not str?

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -1822,10 +1855,21 @@ def _getitem_axis(self, key, axis: int):
...
         # fall thru to straight lookup
         self._validate_key(key, axis)
         return self._get_label(key, axis=axis)
 
+    def _getitem_lower_dim(self, section, key):
+        return getattr(section, self.name)(regex=self.regex)[key]
+
+    def __setitem__(self, key, value):
+        if self.regex:
+            raise TypeError("Inserting with regex not supported")
```
Comment (simonjayhawkins) [inline pandas/core/indexing.py]: should this be `NotImplementedError`?

File: pandas/core/indexing.py (hunk 1 above)
Comment (topper-123) [reply pandas/core/indexing.py]: IMO no.

The .loc interface should anyway accept one axis selector that is not a string (e.g. ``df.loc(regex=True)[df.A ==1, "^col_"]``, so letting neither be a string would not be a big difference and just mean no effect from setting the regex parameter.

File: pandas/core/indexing.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/indexing.py]: Yeah, I will change it.

File: pandas/core/indexing.py [hunk 3]
```diff
@@ -1761,6 +1794,27 @@ def _get_partial_string_timestamp_match_key(self, key, labels):
 
         return key
 
+    def _get_regex_mappings(self, key, axis=None):
+        import re
```
Comment (WillAyd) [inline pandas/core/indexing.py]: Can move this to the top of the module

File: pandas/core/indexing.py (hunk 1 above)
Comment (WillAyd) [reply pandas/core/indexing.py]: I see both points but tend to agree more with @simonjayhawkins. At the very least wouldn't we want to accept other types (bytes, re.pater)?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 1 above)
Comment (topper-123) [reply pandas/core/indexing.py]: Yeah, it could accept bytes. What do you mean by re.pater?

**Comment by jorisvandenbossche** [comment]
I am personally -1 on this proposal, and rather instead would vote for the approach in https://github.com/pandas-dev/pandas/pull/27340 (but using `select`, if that is possible?)

It's of course a personal taste, but those keywords to `loc` are IMO rather unnatural, difficult to discover (you don't typically go looking for the keyword arguments of a `[]` indexing method), and make the behaviour of what is passed to `loc[..]` even more complex.

**Comment by TomAugspurger** [comment]
Agreed with @jorisvandenbossche

**Comment by simonjayhawkins** [comment]
> Agreed with @jorisvandenbossche

OTOH using keywords to extend the loc indexer could make things more consistent as different capabilities are added eg. closed argument https://github.com/pandas-dev/pandas/pull/27209#issuecomment-508780337 instead of adding various different additional methods and apis to extend indexing functionality.

**Comment by WillAyd** [comment]
Would allowing `loc` to natively just handle a regex be an option then? Obviously in other parts of our API we have `regex=True` as a parameter in some function but that seems unnecessary, so if we weren't tied to that we could theoretically implement this without a lot of API additions

**Comment by simonjayhawkins** [comment]
> Would allowing `loc` to natively just handle a regex be an option then?

do you mean allow a regular expression object to be part of the key?

if so, i'd be wary about indexing performance. in #27209 which is just a POC, allowing another type to be part of a key requires additional checks to allow the key to 'pass through' to the appropriate logic.

i liked the suggestion to use a keyword instead, https://github.com/pandas-dev/pandas/pull/27209#issuecomment-508780337 as I think this eliminates the need for these additional checks until the appropriate part of the code is reached and therefore also reducing code complexity (which seemed to be an argument against this approach https://github.com/pandas-dev/pandas/pull/27363#issuecomment-511080544)

**Comment by topper-123** [comment]
I actually agree with @jorisvandenbossche on this.

This PR has been *significantly* more complex to implement than #27340 and will therefore also be more difficult to maintain. This PR required understanding of the class hierarchy of ``_LocIndexer`` and there's always the danger of stepping on other stuff in ``_LocIndexer`` or on its siblings when doing changes here. This will make changing other stuff in .loc/.iloc/.at/.iat more complex also.

#27340 OTOH was very simple to implement and I more-or-less considered it syntactic sugar for the common use case of selecting columns from a regex.

So the approach in #27340 will be better for code maintainability.

@jreback?

**Comment by jreback** [comment]
I think we need to decide how we want to add this functionailty. @topper-123 that is a good data point that this is more complex (which of course we always want to limit as much as possible). OTOH introducing a new method on DataFrame needs a *very* high bar.

So I am -1 on the first for the new method name. I actually like the simplicity of this one. 

All that said. I think this just needs more discussion and consensus.

**Comment by jreback** [comment]
@simonjayhawkins 

> if so, i'd be wary about indexing performance. in #27209 which is just a POC, allowing another type to be part of a key requires additional checks to allow the key to 'pass through' to the appropriate logic.

I am not at all concerned with indexing perf. These are *all* vectorized things; mis-using (say in a loop) is non-idiomatic and not something we really can do much about generally.

**Comment by topper-123** [comment]
To keep the discussion going:

``.loc`` is a general purpose method, so it will be very difficult to add type specific functionality for many axis types.

In that respect I think ``select`` could become the specialized single axis type specific filtering option. So if the user needs something specific to a axis data type (regex here, but it could be other things), they would reach for select, e.g. through an accessor. We would then move type specific subsetting methods from the DataFrame name space, so instead of ``df.at_time``, it would be ``df.select.at_time`` etc.

**Comment by topper-123** [comment]
I've got a simpler implementation for this, so please wait to merge. Will look into it tomorrow.

**Comment by topper-123** [comment]
I've made a simpler implementation, where it doesn't quite feels as if the various parts of .loc/.iloc./.at/.iat step on each toes quite as much. So this version will IMO be much easier to maintain, compared to the previous version.

That makes my worry about adding regex functionality to .loc go away, and I'm neutral whether we add a special ``select_str`` method or add regex functionality to .loc as in this PR (but I do think this should be added one way or another).

**Comment by simonjayhawkins** [comment]
> To keep the discussion going:

just a thought!

a bit out of scope for this PR but may help in deciding on the best course of action.

in this case `regex` is a boolean. so we could also have regex='rows' and regex='columns' if we wanted to limit the scope of regex.

going further off topic, but could still be relevant...

if we also used a similar api for a closed argument, then we wouldn't be able to limit the scope the same way as just suggested.

so would probably require something like `.loc(closed='right', scope='rows')[..., ...]`

but this would still limit the flexibility as the default would only be overridden on one axis.

so maybe `.loc(closed='right', scope='rows')(closed='left', scope='columns')[..., ...]`

starting to get a bit clunky now!

accept a dictionary for closed?

``.loc(closed=dict{'rows':'right', 'columns':'left'})[..., ...]`

so back to regex...

if the scope keyword was added

`.loc(regex=True, closed='right', scope='rows')`

would now be confusing and less flexible. is the scope for closed or regex?

so the dictionary implementation for closed seems more appropriate.

and back to regex...

does being able to limit the scope using regex='columns'/'rows' make sense? not necessarily in this PR though.

> The .loc interface should anyway accept one axis selector that is not a string (e.g. `df.loc(regex=True)[df.A ==1, "^col_"]`, so letting neither be a string would not be a big difference and will ust mean no effect from setting the regex parameter.

could then potentially be ..

`df.loc(regex='columns')[df.A ==1, "^col_"]`

to allow additional validation of the keys.

**Comment by pilkibun** [comment]
This roundabout way of passing arguments `__getitem__` is unidiomatic and I share the [reluctance](https://github.com/pandas-dev/pandas/pull/27363#issuecomment-511080544) that's already been expressed. I do have an interest in adopting a non-intrusive way to extend `loc`. As you know, I want it `loc` to grow a `closed='left'` argument (xref #27060). So here's a new new alternative. **Could we repurpose `loc.__call__` in its entirety and migrate loc into becoming a function?**

```python
df.loc(...)[...] # NO

df.loc[...] # continues to work
df.loc(...) # equivalent
df.loc(..., closed='left') # no longer contentious?
df.loc(..., regex='foo') # good idea or not, but decisionchoice is simplified
```
There might be some impedance mismatch with how `__getitem__` handles its arguments, but I think it should be fairly minimal. This might actually enable cleanups of some ambiguous cases in `loc`.

**Comment by shoyer** [comment]
> **Could we repurpose `loc.__call__` in its entirety and migrate loc into becoming a function?**

I think it would be better to pick a new name for the method. It's too easy to confuse parentheses/brackets, especially for inexperienced Python users.

A new method would also gives the ability to keep-up the API, by using keyword arguments rather than position for selecting on the index or columns, e.g., `df.sel(index='a')` and `df.sel(columns='b')` vs the current `df.loc['a', :]` and `df.loc[:, 'b']`.

**Comment by topper-123** [comment]
A function would drop the ability to slice inputs (``df.loc['a':'c']``), which I think we should keep. It's too late to change fundamental functionality IMO. I'd be ok with new functionality, e.g. ``select`` as I've mentioned previously.

**Comment by pilkibun** [comment]
>  I think it would be better to pick a new name for the method.

But you know I argued for that vigorously and was completely ignored. Do you have a plan for achieving a different outcome? And this `loc()[]` suggestion has divided the maintainers evenly down the middle and seems like another unlikely prospect. How do we move this issue forward?

>  It's too easy to confuse parentheses/brackets

But if they are interchangeable there's little harm, and eventually everyone will migrate to the function form. Much like the `print` statement in python2 only without the pain.

> A new method would also gives the ability to keep-up the API, by using keyword arguments

Yes, but that alternative hasn't had any support. I'm suggesting a third alternative, to transition `loc` into a function, make it backward-compatible with loc, but providing room to grow (`regex`, `closed`). I'm hoping this make it an easier choice for the others to embrace.

**Comment by jreback** [comment]
a function is a nonstarter because you can not assign values; that’s they reason we use getitem/setitem

**Comment by WillAyd** [comment]
Should we punt this one for now and close?

**Comment by WillAyd** [comment]
I think this is stalled so going to close to clean up the queue. @topper-123 feel free to reopen if you disagree

**Comment by topper-123** [comment]
I would like #27617 to get accepted, but that requires that its regex filtering capabilities get a home elsewhere, so the proposal here was to locate it in .loc. I've had another propoasl, where we make a new ``select`` or ``select_str`` method in #27340.

Th underlying question is probably if we're satisfied with the capabilities of ``.filter`` (lots of overlap with other methods, only the regex capability is unique) and that it filters on the axes; I would like the method called ``filter`` to filter on values.

**Pull request closed without merging**